compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.cpp

```diff
@@ -0,0 +1,158 @@
+ *           \               /
+ *          [CircleExpandDims]
+ *                   |
+ *              [CircleNode]
+ *                   |
+ *
+ * AFTER
+ *           |
+ *      [CircleNode]  [CircleConst]
+ *           \              /
+ *           [CircleReshape]
+ *                   |
+ *              [CircleNode]
+ *                   |
+ */
+namespace
+{
+
+int32_t unknown_dim_count(luci::CircleNode *node)
+{
+  int32_t count = 0;
+
+  for (uint32_t i = 0; i < node->rank(); ++i)
+    if (!node->dim(i).known())
+      ++count;
+
+  return count;
+}
+
+/**
+ * @brief Return value in position on CircleConst with int32 format.
+ */
+int32_t value_from_circle_const(const luci::CircleConst *node, uint32_t idx)
+{
+  assert(node->rank() == 1 && node->dim(0).value() > idx);
+  assert(node->dtype() == loco::DataType::S64 || node->dtype() == loco::DataType::S32);
+
+  if (node->dtype() == loco::DataType::S64)
+    return static_cast<int32_t>(node->at<loco::DataType::S64>(idx));
+  return node->at<loco::DataType::S32>(idx);
+}
+
+bool substitute_expand_dims_to_reshape(luci::CircleNode *node)
+{
+  auto target_node = dynamic_cast<luci::CircleExpandDims *>(node);
+  if (target_node == nullptr)
+    return false;
+  auto input_node = loco::must_cast<luci::CircleNode *>(target_node->input());
+  if (input_node->rank() <= 0)
+    return false;
+  auto axis_node = dynamic_cast<luci::CircleConst *>(target_node->axis());
+  if (axis_node == nullptr)
+    return false;
+
+  if (axis_node->dtype() != loco::DataType::S64 && axis_node->dtype() != loco::DataType::S32)
+  {
+    // Abnormal model with unexpected dtype in axis
+    return false;
+  }
+
+  int32_t axis = value_from_circle_const(axis_node, 0);
+  if (axis < 0)
+    // WHY ADD +1?
+    // As minux index is calculated from the last index,
+    // it must expect the new expanded(+1) rank.
+    axis = axis + static_cast<int32_t>(input_node->rank()) + 1;
+
+  auto name = node->name();
+  assert(name.length() > 0);
+
+  auto graph = target_node->graph();
+  auto reshape_node = graph->nodes()->create<luci::CircleReshape>();
+  reshape_node->tensor(input_node);
+  reshape_node->name(name + "/Reshape");
+  luci::add_origin(reshape_node, luci::get_origin(node));
+
+  auto const_node = graph->nodes()->create<luci::CircleConst>();
+  const_node->dtype(loco::DataType::S32);
+  const_node->size<loco::DataType::S32>(input_node->rank() + 1);
+  const_node->shape_status(luci::ShapeStatus::VALID);
+  const_node->rank(1);
+  const_node->dim(0).set(input_node->rank() + 1);
+  for (int32_t i = 0; i < static_cast<int32_t>(input_node->rank()) + 1; i++)
+  {
+    if (i == axis)
+    {
+      const_node->at<loco::DataType::S32>(i) = 1;
+    }
+    else if (i < axis)
+    {
+      const_node->at<loco::DataType::S32>(i) = input_node->dim(i).value();
+    }
+    else
+    {
+      const_node->at<loco::DataType::S32>(i) = input_node->dim(i - 1).value();
+    }
+  }
+  const_node->name(name + "/Reshape/shape");
```

Comment:
Could you add origin to const_node?

Reply:
Does it help that const node having origin, too? 
I took a reference from SubstitutePackToReshapePass, 
- [ ]if so, let me add some code there, too.

Reply:
> Does it help that const node having origin, too?

Yes. It helps to identiy the node name when we do profiling in the backend.